kong/api/routes/upstreams.lua

```diff
@@ -40,7 +42,48 @@ return {
     end,
 
     GET = function(self, dao_factory)
-      crud.paginated_set(self, dao_factory.targets)
+      -- by default, just get a paginated list
+      if not self.params.active then
+        return crud.paginated_set(self, dao_factory.targets)
+      end
+
+      self.params.active = nil
+
+      -- user wanted a list of all active targets for this upstream
+      local target_history, err = dao_factory.targets:find_all({
+        upstream_id = self.params.upstream_id,
+      })
+
+      if not target_history then
+        return app_helpers.yield_error(err)
+      end
+
+      -- sort and walk based on target and creation time
+      for _, target in ipairs(target_history) do
+        target.order = target.target..":"..target.created_at
+      end
+      table.sort(target_history, function(a, b) return a.order > b.order end)
```

Comment:
What happens when you have two sets of two targets with the same create time that look like such:

```
[
    {
      "weight": 0,
      "id": "38b4ab34-f6da-4f88-9e03-aa53f75ed9f7",
      "target": "1.2.3.4:8000",
      "created_at": 1490116516932,
      "upstream_id": "413ea7ee-2740-445d-9c2e-d042e2c7a798"
    },
    {
      "weight": 20,
      "id": "8676ddb9-cd96-40ed-a537-7cc714faac44",
      "target": "1.2.3.4:8000",
      "created_at": 1490116516932,
      "upstream_id": "413ea7ee-2740-445d-9c2e-d042e2c7a798"
    }
]
```

See that `target` and `created_at` are equal; note also the differences in UUIDs (the first element's id is 'lower'). Now consider the following pair:

```
[
    {
      "weight": 0,
      "id": "38b4ab34-f6da-4f88-9e03-aa53f75ed9f7",
      "target": "1.2.3.4:8000",
      "created_at": 1490116516932,
      "upstream_id": "413ea7ee-2740-445d-9c2e-d042e2c7a798"
    }
    {
      "weight": 20,
      "id": "196be764-8fbe-4332-a1d4-c192c7cd6bb5",
      "target": "1.2.3.4:8000",
      "created_at": 1490116516932,
      "upstream_id": "413ea7ee-2740-445d-9c2e-d042e2c7a798"
    }
]
```

So what happens when we try to sort based on `target`, then `created_at`, then `id`? Since it depends solely on `id`, we get different results based on the randomly-defined ID: https://gist.github.com/p0pr0ck5/675d5062f84cf5e04fd311dbb7f36c3f

Now consider the use case of finding all 'active' targets. In such a case, we're essentially looking for the most recent entry in the db for a given row's `target`. As we can see above, it's entirely possible that multiple cases, using the same sort key and method, would produce different results, because the sort key is generated from random, non-deterministic data. This is the _definition_ of an non-deterministic sort. While the behavior is reproducible, it is not useful for a sort :)

Having said all this, it's very much an edge case, and really at this point we're kind of abusing the design of `target` objects (which really aren't objects, per se). So I'm pushing a new version of this patchset with your desired sort key (`target.target..":"..target.created_at..":"..target.id`) so that we can move on :)